lib/ukallocstack/stack.c

```diff
@@ -0,0 +1,277 @@
+	 */
+	UK_ASSERT(vma);
+
+	rc = uk_vma_unmap(sa->vas,
+			  vaddr,
+			  vma->end - vma->start,
+			  0);
+	UK_ASSERT(!rc);
+#else /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+	uk_free(sa->og_a, ptr);
+#endif /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+}
+
+static void *stack_memalign(struct uk_alloc *a, __sz align, __sz size)
+{
+	__vaddr_t sp __maybe_unused;
+	struct uk_stackalloc *sa;
+	int rc __maybe_unused;
+	__vaddr_t vaddr;
+	__sz real_size;
+
+	UK_ASSERT(a);
+	UK_ASSERT(align >= UKARCH_SP_ALIGN);
+	UK_ASSERT(IS_ALIGNED(align,  UKARCH_SP_ALIGN));
+
+	if (!size)
+		return NULL;
+
+	sa = to_stackalloc(a);
+
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	UK_ASSERT(align <= PAGE_SIZE);
+
+	real_size = PAGE_ALIGN_UP(size) + PAGE_SIZE;
+	vaddr = __VADDR_ANY;
+	rc = uk_vma_map_stack(sa->vas,
+			      &vaddr,
+			      real_size,
+			      UK_VMA_MAP_UNINITIALIZED,
+			      NULL,
+			      sa->initial_size);
+	if (unlikely(rc)) {
+		uk_pr_err("Failed to create stack VMA of size 0x%lx\n",
+			  real_size);
+		return NULL;
+	}
+#else /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+	real_size = size;
+	vaddr = (__vaddr_t)uk_memalign(sa->og_a, align, real_size);
+	if (unlikely(!vaddr)) {
+		uk_pr_err("Failed to allocate stack of sizegth 0x%lx from "
+			  "allocator %p\n", real_size, sa->og_a);
+		return NULL;
+	}
+#if CONFIG_LIBUKVMEM
+	sp = ukarch_gen_sp(vaddr, real_size);
+	rc = uk_vma_advise(sa->vas,
+			   PAGE_ALIGN_DOWN((__vaddr_t)
+					   (PAGE_ALIGN_UP(sp) -
+					    PAGE_ALIGN_UP(sa->initial_size))),
+			   PAGE_ALIGN_UP(sa->initial_size),
+			   UK_VMA_ADV_WILLNEED,
+			   UK_VMA_FLAG_UNINITIALIZED);
+	if (unlikely(rc)) {
+		uk_pr_err("Could not pre-map stack initial size\n");
+		return NULL;
+	}
+#endif /* CONFIG_LIBUKVMEM */
+#endif /* !CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+
+	return (void *)vaddr;
+}
+
+static int stack_posix_memalign(struct uk_alloc *a, void **memptr, __sz align,
+				__sz size)
+{
+	void *addr;
+
+	UK_ASSERT(a);
+	UK_ASSERT(memptr);
+	UK_ASSERT(align >= UKARCH_SP_ALIGN);
+	UK_ASSERT(IS_ALIGNED(align,  UKARCH_SP_ALIGN));
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	UK_ASSERT(align <= PAGE_SIZE);
+#endif /* CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
+
+	addr = stack_memalign(a, align, size);
+	if (unlikely(!addr)) {
+		*memptr = NULL;
+		return -ENOMEM;
+	}
+
+	*memptr = addr;
+	return 0;
+}
+
+static void *stack_calloc(struct uk_alloc *a, __sz nmemb, __sz size)
+{
+	__sz total;
+	void *addr;
+
+	UK_ASSERT(a);
+
+	if (nmemb <= 1 || !size)
+		return NULL;
+
+	total = nmemb * size;
+	addr = stack_malloc(a, total);
+	if (unlikely(!addr))
+		return NULL;
+
+	memset(addr, 0, total);
+
+	return addr;
+}
+
+struct uk_alloc *uk_allocstack_init(struct uk_alloc *a,
+#if CONFIG_LIBUKVMEM
+				     struct uk_vas *vas,
+				     __sz initial_size
+#endif /* CONFIG_LIBUKVMEM */
+				     )
+{
+	struct uk_stackalloc *sa;
+
+	UK_ASSERT(a);
+#if CONFIG_LIBUKVMEM
+	UK_ASSERT(vas);
+#endif /* CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
```

Comment:
Note: just noticed I have lots of `CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS` instead of `CONFIG_LIBUKVMEM` comments.